### ISSUE TITLE
populateFormValues does not update values of incomplete multi-pages form

### ISSUE THREAD

**Issue by romainpoirier** [verbb/formie#658]
**Description**
I have a multi-pages forms where the user can pause and complete later its form. Some hidden fields of this form are populated by values that can be updated from another source. Example: a User field, or an Entry field.

On init, the form is populated from the values it gets. But when going forward or back, these values are never updated when the populated data are.

In my memories, this was previously working as expected. Maybe this is linked to the `Fixed Hidden fields not always having the default value set on submission.` bug fixed in 1.4.24 release?

Below is a complete example:
```
{% set submission = craft.formie.submissions.form(handle).user(currentUser).isIncomplete(true).one() %}

{% set populateFormValues = {
	myField: 'myValue'
	companyName: currentUser.companyName,
	entryTitle: currentUser.myField.one().title
} %}

{% set serializedFieldValues = {} %}
{% if submission is not empty %}
	{% for handle, value in submission.getSerializedFieldValuesForIntegration() %}
		{% set serializedFieldValues = serializedFieldValues | merge({
			(handle): (populateFormValues[handle] ?? null is not empty and populateFormValues[handle] != value) ? populateFormValues[handle] : value
		}) %}
	{% endfor %}
{% endif %}

{% if submission is not empty %}
	{% do craft.formie.populateFormValues(submission.form, serializedFieldValues) %}
	{% do craft.formie.setCurrentSubmission(form, submission) %}
{% else %}
	{% do craft.formie.populateFormValues(form, populateFormValues) %}
	{% set submission = craft.formie.submissions.form(handle).user(currentUser).isIncomplete(false).one() %}
{% endif %}

<!-- Test: -->
{{ d({
	serializedFieldValues: serializedFieldValues.myField ?? null,
	getSerializedFieldValuesForIntegration: submission.getSerializedFieldValuesForIntegration().myField ?? null,
}) }}
```

`serializedFieldValues.myField ` get the correct updated value. `submission.getSerializedFieldValuesForIntegration().myField` get the first value when starting the form and is never updated later when editing the form (before submitting it).

### Form settings
- Multi-page form: Yes
- Submission Method: Page Reload
- Client-side Validation: Yes
- Custom Form Templates: Yes

**Additional info**
- Plugin version: 1.4.27
- Craft version: 3.7.23
- Multi-site: Yes

**Comment by engram-design** [comment]
This is somewhat of a known issue I believe, because the `populateFormValues()` only sets the default values of fields, as opposed to overriding values. Otherwise, we'd just continually be forcing the value of a field, overriding anything the user inputs.

From what I've found, this is really only an issue when you're testing, and you're on say page 2 of a form, then add your `populateFormValues()` calls - nothing will happen, because you're trying to populate an incomplete submission. You'd need to restart the submission again so the population event would apply default values.

If the same scenario for you, if you set a default value on page 2 for example, when the user submits the first page, this value will be set on the submission. The only way this won't occur is for any submissions that exist before you add your `populateFormValues()` code - values won't retroactively be applied.

But yes, this is a good point in your scenario where users are coming back to a saved form, and data is changing between their visits.

Now while we could do checks to see if the field value is "empty" and if so, apply the default from `populateFormValues()`, but that gets tricky. What if your user has intentionally made that value empty, removing the default setup? This probably doesn't make sense for an element-based field, but think something like a Single-Line Text field, where you might have "Enter any optional text here" as the default set in `populateFormValues()`, but the user removes this text. With our logic, it would populate the value again, against the users intentions.

I'm not sure if it's ever worked (I might be wrong), but I'll give it some thought.

**Comment by romainpoirier** [comment]
`This is somewhat of a known issue I believe, because the populateFormValues() only sets the default values of fields, as opposed to overriding values.`
Is there another function to send one single object of data to override and update at each step of an incomplete multi-pages form?

`this is really only an issue when you're testing`
Unfortunately this is a real live issue for my project, and I'm afraid it occurs in production while I was thinking it works as expected before...

`Otherwise, we'd just continually be forcing the value of a field, overriding anything the user inputs.`
`What if your user has intentionally made that value empty, removing the default setup? `
In my case, I'm using `populateFormValues()` to only fill hidden fields, so that's the expected result.
I need to always force value to the latest up to date value. The user won't see these fields, he can't change their values.

**Comment by engram-design** [comment]
Hmm, I might suggest I add a "force" option to that setting, because forcing values to _always_ have a particular value isn't always desirable. And when I mean testing, I mean when you're testing using the `populateFormValues()` and adding values during development where you're changing your templates, but still have an incomplete submission you're using. I can see this is an issue in your case though.

Even if I were to add the forced behaviour for just hidden fields, that gets tricky when the values might be updated through other means (JS, possibly others). That's probably why I'm thinking of adding a "force" option to `populateFormValues()` or an entirely different function like `setFormValues()`.

**Comment by romainpoirier** [comment]
Thank you, the `populateFormValues()` with `force` option or `setFormValues()` would be super nice. In my case, I'm using Page Reload as Submission Method, so it would works as expected.

**Comment by engram-design** [comment]
Just added the force parameter - https://verbb.io/craft-plugins/formie/docs/template-guides/populating-forms#forcing-values

To get this early, change your `verbb/formie` requirement in `composer.json` to:

```
"require": {
  "verbb/formie": "dev-craft-3 as 1.4.27",
  "...": "..."
}
```

Then run `composer update`.

**Comment by romainpoirier** [comment]
Thank you!

I have updated my initial code example to this:
```
{% do craft.formie.populateFormValues(submission.form, serializedFieldValues, true) %}
{% do craft.formie.populateFormValues(form, populateFormValues, true) %}
```
Unfortunately, in my case it change nothing, the data are still not updated.
When I print the `submission.myField` value, it still get the initial one.
I tried with both existing and new multi-steps incomplete submissions.

Any idea why?

**Comment by engram-design** [comment]
Any reason for the multiple calls? Or are you just showing the two things you've tried? What are `serializedFieldValues` and `populateFormValues` variables?

**Comment by romainpoirier** [comment]
Sorry if I wasn't clear. There's only one call, within a condition. My previous code was an extract of what I edited. Here, in context:

```
{% if submission is not empty %}
	{% do craft.formie.populateFormValues(submission.form, serializedFieldValues, true) %}
	{% do craft.formie.setCurrentSubmission(form, submission) %}
{% else %}
	{% do craft.formie.populateFormValues(form, populateFormValues, true) %}
	{% set submission = craft.formie.submissions.form(handle).user(currentUser).isIncomplete(false).one() %}
{% endif %}
```

I have a condition to check if the incomplete submission exists. If it exists, I have to set it using `setCurrentSubmission` in order to edit it without creating a new one.

`populateFormValues` is an object of all hidden fields I have to fill when the submission is created.
`serializedFieldValues` is like `populateFormValues`, but will construct a clean object of what should be filled at the creation of the form and what is currently filled in the form, without erasing both of this data:

```
{% set populateFormValues = {
	myField: 'myValue'
	companyName: currentUser.companyName,
	entryTitle: currentUser.myField.one().title
} %}

{% set serializedFieldValues = {} %}
{% if submission is not empty %}
	{% for handle, value in submission.getSerializedFieldValuesForIntegration() %}
		{% set serializedFieldValues = serializedFieldValues | merge({
			(handle): (populateFormValues[handle] ?? null is not empty and populateFormValues[handle] != value) ? populateFormValues[handle] : value
		}) %}
	{% endfor %}
{% endif %}
```

**Comment by engram-design** [comment]
Ah, I see, the combination of the `setCurrentSubmission()` will throw things off, because the submission data won't persist, because you're setting it on the form's submission, not the submission you're fetching beforehand and using in `setCurrentSubmission()`. Not only that but `setCurrentSubmission()` does some things to prep the element with it's content.

I've added the possibility to pass in a form or submission to `populateFormValues()`. This will be the first step, instead of using `submission.form` you'll want to use `submission` (when one exists). But you'll want to call `setCurrentSubmission()` first.

```
{% set form = craft.formie.forms({ handle: 'myForm' }).one() %}
{% set submission = craft.formie.submissions.isIncomplete(true).form(form).one() %}

{% do craft.formie.setCurrentSubmission(form, submission) %}

{% do craft.formie.populateFormValues(submission, {
    myField: 'myValue',
}, true) %}

{{ craft.formie.renderForm(form) }}
```

Or, thinking about this further, you could even just set the field values directly - which is how it would be done natively (for any element), without using `populateFormValues()` entirely.

```
{% set form = craft.formie.forms({ handle: 'myForm' }).one() %}
{% set submission = craft.formie.submissions.isIncomplete(true).form(form).one() %}

{% do craft.formie.setCurrentSubmission(form, submission) %}

{% do submission.setFieldValues({
    myField: 'myValue',
}, true) %}

{{ craft.formie.renderForm(form) }}
```

**Comment by romainpoirier** [comment]
Wonderful! Thank you for your patience and explanations. This now works as expected.

**Comment by engram-design** [comment]
Fixed in [1.4.28](https://github.com/verbb/formie/releases/tag/1.4.28)
